port/zos390/omrvmem.c

```diff
@@ -1296,29 +1299,40 @@ omrvmem_default_large_page_size_ex(struct OMRPortLibrary *portLibrary, uintptr_t
 	}
 #endif /* OMR_ENV_DATA64 */
 
-	while (0 != PPG_vmem_pageSize[index]) {
-		if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]))
-			) {
-				break;
+
+	/* Determine maximum pageable, and fixed large pages available. */
+	for (int i = 0; ((0 != PPG_vmem_pageSize[i]) && (ONE_M >= PPG_vmem_pageSize[i])); i++) {
+
+		if (ONE_M == PPG_vmem_pageSize[i]) {
+
+			if (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[i])) {
+				/* Pageable. Keep track of largest pageable size */
+				indexPageable = i;
 			}
 #if defined(OMR_ENV_DATA64)
-		} else {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index]))
-			) {
-				break;
+			else if (0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[i])) {
+				/* Fixed. Keep track of largest fixed size */
+				index = i;
 			}
-#endif
+#endif /* OMR_ENV_DATA64 */
 		}
-		index++;
 	}
+
+	/* 
+	 * When mode is OMRPORT_VMEM_MEMORY_EXECUTE we will only consider pagable large pages.
+	 * When flag is OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE, we prefer pageable unless fixed is larger.
+	 * NOTE: 32 bit only supports Pageable large pages. OMRPORT_VMEM_MEMORY_MODE_EXECUTE is guaranteed to be high during 32bit execution.
+	 */
+	if ((0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) ||
+	   ((NULL != pageFlags) && (-1 != indexPageable) && (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE & (*pageFlags))))) {
+		index = indexPageable;
```

Comment:
I believe this is not correct. If `Pageable` not provisioned `indexPageable` would be -1. And if `OMRPORT_VMEM_MEMORY_MODE_EXECUTE` is requested you are going set this -1 to `index` and use it as array index.
You should seriously work on expressing your thoughts in the code. An implementation should be readable. An in this function for example you replace relatively simple readable code to absolutely terrible. And you caught yourself and did this mistake.
Your intention here is to say: "Ok, pageable is provisioned. So use it for Executable mode or if Preferable Pageable flag is set. Also  `pageFlags` might be NULL potentially. And  Preferable Pageable flag can not be provided if `(NULL == pageFlags)`
You can rearrange code:
```
BOOLEAN pageablePreferableRequested = (NULL != pageFlags) && (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE & (*pageFlags));
......
    if ((-1 != indexPageable) && ((0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) || pageablePreferableRequested)) {
        index = indexPageable;
    }
```
`pageablePreferableRequested` can be used to simplify line 1292 as well

Reply:
> I believe this is not correct. If Pageable not provisioned indexPageable would be -1. And if OMRPORT_VMEM_MEMORY_MODE_EXECUTE is requested you are going set this -1 to index and use it as array index.

Let's take a step back. 
OMRPORT_VMEM_MEMORY_MODE_EXECUTE is only requested by the JIT Codecache. Which does NOT support fixed large pages.

Have you looked at the code fully? There's no way we're accessing a index out of range -1. 0 would be returned as the pagesize, and flags if a pageable 1M page was not provisioned. In this case the codecache would continue to use 4K pageable which is always provisioned on the system. 

>  Also pageFlags might be NULL potentially.
That's already considered.

Reply:
I apologize if my comment sounded offensive. It was not an intention.
I still think there is a mistake in this line and it was overseen because an implementation of the code is not readable enough even for code author.

Reply:
Firstly, what you've requested does not really address any real functional issue. You're suggesting that we simply store whether `pageablePreferable` is requested in a variable. Which is a respectable contribution and I agree that would simplify the code. **_I agree to make that change._**

I must point out that there is something functionally incorrect with the second half of your contribution. We need to set the index to -1 if `OMRPORT_VMEM_MEMORY_MODE_EXECUTE` mode is used, and 1M pageable large pages are not configured on the system. As I've stated before, the codecache is the only component that uses executable large pages. The Z/OS codecache does not support fixed large pages, and we absolutely can not return the fixed large page as the default. Considering what I've just stated, a valid change would look like this:

```
if ((0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) || ((pageablePreferableRequested) && (-1 != indexPageable)) {
        index = indexPageable;
    }
```

Be assured, this has passed J9vmemtest, and all relevant internal and open tests. Let me know if you actually find something functionally incorrect. 

I also suggest you take a look at the relevant community guidelines. I look forward to continue a constructive discussion.
https://www.eclipse.org/org/documents/Community_Code_of_Conduct.php
https://help.github.com/en/github/site-policy/github-community-guidelines

Reply:
See 98f388ec1a11c49b391cc28de8fb9a7953139bfd

Reply:
Now, when you fixed bug by checking for -1 before set `*pageSize` and `*pageFlags` I insist that there is no need to go inside this if in case if `OMRPORT_VMEM_MEMORY_MODE_EXECUTE` requested but `indexPageable` still be -1

Reply:
ok, last comment was not accurate because in case of `OMRPORT_VMEM_MEMORY_MODE_EXECUTE` overwrite `index` by  `indexPageable` even it is -1. Would you please comment in this place to prevent confusion in future?

Reply:
Yes, if the mode is `OMRPORT_VMEM_MEMORY_MODE_EXECUTE` we need to overwrite the index to `indexPageable`, even if  `indexPageable` is `-1`. This is because index may be pointing to a fixed large page, and we don't want to return a fixed large page under that mode.